### PR TITLE
feat(application-system): Add validated config

### PR DESCRIPTION
**Pull request by norda-gunni**
# ...

Attach a link to issue if relevant

## What

Specify what you're trying to achieve

## Why

Specify why you need to achieve this

## Screenshots / Gifs

Attach Screenshots / Gifs to help reviewers understand the scope of the pull request

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Formatting passes locally with my changes
- [ ] I have rebased against main before asking for a review

## Summary by CodeRabbit

- **New Features**
	- Introduced a shared module configuration to enhance modularization and reuse across the application.
	- Added a robust configuration management system for the Template API, ensuring all necessary parameters are validated and loaded correctly.
	- Added error handling for user profile data retrieval, improving user experience.
	- Introduced a new exported constant `dynamicModules` for improved module management.

- **Bug Fixes**
	- Improved test configurations to utilize shared module settings, enhancing the reliability of test scenarios.

- **Refactor**
	- Streamlined module definitions by replacing dynamic registration methods with decorator-based approaches for improved clarity and maintainability.
	- Updated configuration access methods across various services to utilize the new shared module configuration, enhancing modularity and code clarity.

### CODE REVIEW

File: apps/application-system/api/src/app/modules/application/e2e/application.spec.ts [hunk 1]
```diff
@@ -56,6 +57,17 @@ class MockContentfulRepository {
   }
 }
 
+const mockConfig = {
+  jwtSecret: 'supersecret',
+  clientLocationOrigin: 'http://localhost:4200',
```
Comment (Toti91) [inline apps/application-system/api/src/app/modules/application/e2e/application.spec.ts:62]: If this is for the application-system frontend, then the port should be `4242` no?

**Review by Toti91 — COMMENTED** [review]
One comment, otherwise I think @obmagnusson should review this also

File: libs/application/template-api-modules/src/lib/modules/shared/shared.module.ts [hunk 2]
```diff
@@ -20,7 +21,7 @@ export class SharedTemplateAPIModule {
       module: SharedTemplateAPIModule,
       imports: [
         ConfigModule.forRoot({
-          load: [configuration],
+          load: [configuration, sharedModuleConfig],
```
Comment (obmagnusson) [inline libs/application/template-api-modules/src/lib/modules/shared/shared.module.ts:24]: Remove? Should be enough to load this in the app.module. 
We should also get of the dynamic registers of the SharedTemplateApiModule

File: libs/application/template-api-modules/src/lib/modules/template-api.module.ts [hunk 3]
```diff
@@ -9,12 +9,17 @@ import { BaseTemplateAPIModuleConfig } from '../types'
...
+import { templateAPIConfig } from './template-api.config'
+import { ConfigModule } from '@nestjs/config'
 
 export class TemplateAPIModule {
   static register(config: BaseTemplateAPIModuleConfig): DynamicModule {
     return {
       module: TemplateAPIModule,
       imports: [
+        ConfigModule.forRoot({
+          load: [templateAPIConfig],
```
Comment (obmagnusson) [inline libs/application/template-api-modules/src/lib/modules/template-api.module.ts:21]: should just be loaded in the app.module

**Review by obmagnusson — CHANGES_REQUESTED** [review]

File: libs/application/template-api-modules/src/lib/modules/shared/shared.module.ts (hunk 2 above)
Comment (norda-gunni) [reply libs/application/template-api-modules/src/lib/modules/shared/shared.module.ts:24]: Yup, thanks.

File: libs/application/template-api-modules/src/lib/modules/template-api.module.ts (hunk 3 above)
Comment (norda-gunni) [reply libs/application/template-api-modules/src/lib/modules/template-api.module.ts:21]: Thanks for the insight 👍

File: libs/application/template-api-modules/src/lib/modules/template-api.module.ts [hunk 4]
```diff
@@ -9,6 +9,8 @@ import { BaseTemplateAPIModuleConfig } from '../types'
 import { TemplateAPIService } from './template-api.service'
 import { TEMPLATE_API_SERVICES } from './template-api.constants'
 import { LoggingModule } from '@island.is/logging'
+import { templateAPIConfig } from './template-api.config'
+import { ConfigModule } from '@nestjs/config'
```
Comment (obmagnusson) [inline libs/application/template-api-modules/src/lib/modules/template-api.module.ts:13]: these 2 are unused.

File: libs/application/template-api-modules/src/lib/modules/shared/shared.module.ts [hunk 5]
```diff
@@ -11,6 +11,7 @@ import { SharedTemplateApiService } from './shared.service'
 import { SmsModule } from '@island.is/nova-sms'
 import { PaymentModule } from '@island.is/application/api/payment'
 import { AttachmentS3Service } from './services'
+import { sharedModuleConfig } from './shared.config'
```
Comment (obmagnusson) [inline libs/application/template-api-modules/src/lib/modules/shared/shared.module.ts:14]: unused

File: libs/application/template-api-modules/src/lib/modules/shared/shared.module.ts [hunk 6]
```diff
@@ -11,6 +11,7 @@ import { SharedTemplateApiService } from './shared.service'
...
```
Comment (obmagnusson) [inline libs/application/template-api-modules/src/lib/modules/shared/shared.module.ts:15]: Should make this one use the shared confing and remove the dynamic register

**Review by obmagnusson — CHANGES_REQUESTED** [review]

**Review by ylfahfa — APPROVED** [review]

**Review by HjorturJ — APPROVED** [review]

**Review by thorkellmani — APPROVED** [review]
LGTM

**Review by rafnarnason — APPROVED** [review]

**Review by FridrikMProgramm — APPROVED** [review]

File: libs/application/template-api-modules/README.md [hunk 7]
```diff
@@ -35,7 +35,7 @@ export class ReferenceTemplateModule {
   static register(config: BaseTemplateAPIModuleConfig): DynamicModule {
```
Comment (obmagnusson) [inline libs/application/template-api-modules/README.md:35]: We should use the opportunity to refactor these modules from DynamicModule and remove the register function as it is no longer needed.

File: apps/application-system/api/src/app/modules/application/application.module.ts [hunk 8]
```diff
@@ -35,10 +34,7 @@ import { ApplicationActionService } from './application-action.service'
     PaymentModule,
     AuditModule.forRoot(environment.audit),
     AuthModule.register(environment.auth),
-    TemplateAPIModule.register({
-      ...environment.templateApi,
-      applicationService: TemplateApiApplicationService,
-    }),
+    TemplateAPIModule.register(environment.templateApi),
```
Comment (obmagnusson) [inline apps/application-system/api/src/app/modules/application/application.module.ts:37]: This should not be needed any more and The TemplateAPIModule should not be Dynamic anymore.

**Review by obmagnusson — CHANGES_REQUESTED** [review]
Nice work! Just one small step that we maybe should take. Also cant we get rid of the Environments folder?

File: libs/application/template-api-modules/README.md (hunk 7 above)
Comment (norda-gunni) [reply libs/application/template-api-modules/README.md]: done

File: apps/application-system/api/src/app/modules/application/application.module.ts (hunk 8 above)
Comment (norda-gunni) [reply apps/application-system/api/src/app/modules/application/application.module.ts:37]: done

**Review by obmagnusson — APPROVED** [review]

**Review by jonnigs — APPROVED** [review]
LGTM!

File: libs/application/template-api-modules/src/lib/modules/shared/shared.config.ts [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+      cert: z.string(),
+      passphrase: z.string(),
+      basePath: z.string(),
+    }),
+  }),
+  contentful: z.object({
+    accessToken: z.string(),
+  }),
+})
+
```
[the remark this answers is not in the document]
Comment (norda-gunni) [reply libs/application/template-api-modules/src/lib/modules/shared/shared.config.ts:46]: Done

**Pull request merged**